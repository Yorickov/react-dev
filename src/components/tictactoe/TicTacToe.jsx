import React, { useState } from 'react';
import Board from './Board.jsx';
import { calculateWinner } from '../../lib/utils';

const TicTacToe = () => {
  const [history, setHistory] = useState([{
    squares: Array(9).fill(null),
  }]);
  const [xIsNext, setSign] = useState(true);
  const [stepNumber, setStep] = useState(0);

  const handleClick = (i) => () => {
    const actualHistory = history.slice(0, stepNumber + 1);
    const { squares } = actualHistory[actualHistory.length - 1];
    if (calculateWinner(squares) || squares[i]) {
      return;
    }

    const sign = xIsNext ? 'X' : 'O';
    const newSquares = [
      ...squares.slice(0, i), sign, ...squares.slice(i + 1),
    ];

    setHistory([...actualHistory, { squares: newSquares }]);
    setStep(actualHistory.length);
    setSign(!xIsNext);
  };

  const jumpTo = (step) => {
    setStep(step);
    setSign((step % 2) === 0);
  };

  const moves = history.map((step, move) => {
    const description = (move === 0) ? 'To the start' : `Move to step #${move}`;
    return (
      <li key={move}>
        <button onClick={jumpTo.bind(null, move)}>{description}</button>
      </li>
    );
  });

  const { squares } = history[stepNumber];
  const winner = calculateWinner(squares);
  const status = winner ? (`The winner is ${winner}`)
    : `Next step: ${xIsNext ? 'X' : 'O'}`;

  return (
    <div className="row">
      <div className="col-auto game-board">
        <Board
          squares={squares}
          onClick={handleClick}
        />
      </div>
      <div className="col-8 game-info">
        <div className='status'>{status}</div>
        <ol>{moves}</ol>
      </div>
    </div>
  );
};

export default TicTacToe;
