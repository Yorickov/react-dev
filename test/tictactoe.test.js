import React from 'react';
import TicTacToe from '../src/components/tictactoe/TicTacToe.jsx';

it('TicTacToe', () => {
  const wrapper = mount(<TicTacToe />);

  expect(wrapper.find('.game-info li')).toHaveLength(1);
  expect(wrapper.find('.game-info .status').text()).toEqual('Next step: X');
  expect(wrapper.render()).toMatchSnapshot();

  const firstLine = wrapper.find('.table tr').at(0);
  const secondLine = wrapper.find('.table tr').at(1);

  firstLine.find('button').at(0).simulate('click');
  expect(wrapper.find('.game-info .status').text()).toEqual('Next step: O');
  secondLine.find('button').at(0).simulate('click');

  firstLine.find('button').at(1).simulate('click');
  secondLine.find('button').at(1).simulate('click');

  firstLine.find('button').at(2).simulate('click');
  expect(wrapper.find('.game-info li')).toHaveLength(6);
  expect(wrapper.find('.game-info .status').text()).toEqual('The winner is X');

  expect(wrapper.render()).toMatchSnapshot();
});
