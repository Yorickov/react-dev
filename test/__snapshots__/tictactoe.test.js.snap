// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicTacToe 1`] = `
<div
  class="row"
>
  <div
    class="col-auto game-board"
  >
    <table
      class="table table-bordered"
    >
      <tbody>
        <tr>
          <td>
            <button
              class="square border-0"
            />
          </td>
          <td>
            <button
              class="square border-0"
            />
          </td>
          <td>
            <button
              class="square border-0"
            />
          </td>
        </tr>
        <tr>
          <td>
            <button
              class="square border-0"
            />
          </td>
          <td>
            <button
              class="square border-0"
            />
          </td>
          <td>
            <button
              class="square border-0"
            />
          </td>
        </tr>
        <tr>
          <td>
            <button
              class="square border-0"
            />
          </td>
          <td>
            <button
              class="square border-0"
            />
          </td>
          <td>
            <button
              class="square border-0"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="col-8 game-info"
  >
    <div
      class="status"
    >
      Next step: X
    </div>
    <ol>
      <li>
        <button>
          To the start
        </button>
      </li>
    </ol>
  </div>
</div>
`;

exports[`TicTacToe 2`] = `
<div
  class="row"
>
  <div
    class="col-auto game-board"
  >
    <table
      class="table table-bordered"
    >
      <tbody>
        <tr>
          <td>
            <button
              class="square border-0"
            >
              X
            </button>
          </td>
          <td>
            <button
              class="square border-0"
            >
              X
            </button>
          </td>
          <td>
            <button
              class="square border-0"
            >
              X
            </button>
          </td>
        </tr>
        <tr>
          <td>
            <button
              class="square border-0"
            >
              O
            </button>
          </td>
          <td>
            <button
              class="square border-0"
            >
              O
            </button>
          </td>
          <td>
            <button
              class="square border-0"
            />
          </td>
        </tr>
        <tr>
          <td>
            <button
              class="square border-0"
            />
          </td>
          <td>
            <button
              class="square border-0"
            />
          </td>
          <td>
            <button
              class="square border-0"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="col-8 game-info"
  >
    <div
      class="status"
    >
      The winner is X
    </div>
    <ol>
      <li>
        <button>
          To the start
        </button>
      </li>
      <li>
        <button>
          Move to step #1
        </button>
      </li>
      <li>
        <button>
          Move to step #2
        </button>
      </li>
      <li>
        <button>
          Move to step #3
        </button>
      </li>
      <li>
        <button>
          Move to step #4
        </button>
      </li>
      <li>
        <button>
          Move to step #5
        </button>
      </li>
    </ol>
  </div>
</div>
`;
